# 00. Table of content

### 01. Importing Libraries & Dataframes
### 02. Create Subset
### 03. Grouping Data with Pandas
### 04. Multiple Aggregrations
### 05. Aggregating Data with transform()
### 06. Adjust Data Types and Drop Columns before Proceeding with Tasks
### 07. Group Entire df by Department_id and Aggregate by Order_number
### 08. Difference between Spending Habits of the Three Customer Types
### 09. Identify Ddifferent Types of Spenders for a Marketing Campaign
### 10. Determine frequent vs. non-frequent Customers
### 11. Exporting Dataframes




# 01. Importing Libraries & Dataframes

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create path
path = r'C:\Users\krist\22.01.2024 Instacart Basket Analysis'

In [3]:
# Import last dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

# 02. Create Subset

In [4]:
# Create subset of df_merge
df = ords_prods_merge[:1000000]

In [5]:
df.shape

(1000000, 18)

In [6]:
df.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

      busiest_day    busiest_days busiest_period_of_day  
0  regularly busy  regularly busy        average orders  
1  regularly busy      least busy        average orders  
2  regularly busy      least busy           most orders  
3      least busy      least busy        average orders  
4      least busy      least busy           most orders

# 03. Grouping Data with Pandas

In [7]:
# Group by product_name
df.groupby('product_name')

In [8]:
# Group by department_id and perform a single aggregation
df.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
# Alternative command to group and aggregate 
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

# 04. Multiple Aggregations

In [10]:
# Group by department_id and perform multiple aggregation
result = ords_prods_merge.groupby('department_id').agg({'order_number':['mean', 'min', 'max','count']})

In [11]:
result_sorted = result.sort_values(('order_number', 'count'), ascending=False)

In [12]:
result_sorted

order_number                 
                      mean min max    count
department_id                              
4                17.811403   1  99  9479291
16               17.665606   1  99  5398747
19               17.177343   1  99  2887550
7                17.225802   1  99  2688123
1                15.457838   1  99  2234743
13               16.583536   1  99  1875369
3                17.170395   1  99  1172428
15               16.165037   1  99  1068058
20               16.473447   1  99  1051249
9                15.895474   1  99   866627
17               15.694469   1  99   738666
12               15.887671   1  99   708927
14               16.773669   1  99   703033
11               16.170638   1  99   447572
18               19.310397   1  99   423802
6                16.439806   1  99   269253
5                15.215751   1  99   153696
8                15.340650   1  99    97716
21               22.902379   1  99    69145
2                17.277920   1  99    36291
10               20.197148   1  99    34573

In [13]:
result_sorted.to_clipboard()

# 05. Aggregating Data with transform()

In [14]:
# Locate loyalty customers
# Group by user_id and transform order_number to max
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\krist\AppData\Local\Temp\ipykernel_13288\3454195081.py:3: FutureWarning: The provided callable <function amax at 0x0000015005C21A80> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [15]:
# Check on user_id 1 --> 10 orders matches max_order of 10
ords_prods_merge[ords_prods_merge['user_id'] == 1]

order_id  user_id eval_set  order_number  order_day_of_week  \
0        2539329        1    prior             1                  2   
1        2398795        1    prior             2                  3   
2         473747        1    prior             3                  3   
3        2254736        1    prior             4                  4   
4         431534        1    prior             5                  4   
5        3367565        1    prior             6                  2   
6         550135        1    prior             7                  1   
7        3108588        1    prior             8                  1   
8        2295261        1    prior             9                  1   
9        2550362        1    prior            10                  4   
35791    2539329        1    prior             1                  2   
51726    2539329        1    prior             1                  2   
51727    2398795        1    prior             2                  3   
51728     473747        1    prior             3                  3   
51729    2254736        1    prior             4                  4   
51730     431534        1    prior             5                  4   
51731    3367565        1    prior             6                  2   
51732     550135        1    prior             7                  1   
51733    3108588        1    prior             8                  1   
51734    2295261        1    prior             9                  1   
51735    2550362        1    prior            10                  4   
58202    2539329        1    prior             1                  2   
58203    2398795        1    prior             2                  3   
60725    2539329        1    prior             1                  2   
60726    2254736        1    prior             4                  4   
61939    2398795        1    prior             2                  3   
61940     473747        1    prior             3                  3   
61941    2254736        1    prior             4                  4   
61942     431534        1    prior             5                  4   
61943    3367565        1    prior             6                  2   
61944     550135        1    prior             7                  1   
61945    3108588        1    prior             8                  1   
61946    2295261        1    prior             9                  1   
61947    2550362        1    prior            10                  4   
63885    2398795        1    prior             2                  3   
63886     431534        1    prior             5                  4   
443335   2398795        1    prior             2                  3   
443336    550135        1    prior             7                  1   
443337   2550362        1    prior            10                  4   
447086    473747        1    prior             3                  3   
447087   2254736        1    prior             4                  4   
447088    431534        1    prior             5                  4   
447089   3367565        1    prior             6                  2   
447090    550135        1    prior             7                  1   
447091   3108588        1    prior             8                  1   
447092   2295261        1    prior             9                  1   
447093   2550362        1    prior            10                  4   
453282    473747        1    prior             3                  3   
474772    431534        1    prior             5                  4   
480298    431534        1    prior             5                  4   
494178    431534        1    prior             5                  4   
529591   3108588        1    prior             8                  1   
529592   2295261        1    prior             9                  1   
529593   2550362        1    prior            10                  4   
538149   3108588        1    prior             8                  1   
538150   2295261        1    pri

In [16]:
# Check via head function
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
..       ...      ...      ...           ...                ...   
95   3226575      360    prior             1                  5   
96   1469869      377    prior             3                  5   
97   1927023      387    prior             2                  4   
98    858092      420    prior             4                  1   
99    119314      420    prior             5                  3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     NaN         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered product_name  aisle_id  department_id  prices _merge  \
0           0         Soda        77              7     9.0   both   
1           1         Soda        77              7     9.0   both   
2           1         Soda        77              7     9.0   both   
3           1         Soda        77              7     9.0   both   
4           1         Soda        77              7     9.0   both   
..        ...          ...       ...            ...     ...    ...   
95          0         Soda        77              7     9.0   both   
96          0         Soda        77              7     9.0   both   
97          0         Soda        77              7     9.0   both   
98          0         Soda        77              7     9.0   both   
99          1         Soda        77              7     9.0   both   

       busiest_day    busiest_days busiest_period_of_day  max_order  
0   regularly busy  regularly busy        average orders         10  
1   regularly busy      least busy        average orders         10  
2   regularly busy      least busy           most orders         10  
3       least busy      least busy        average orders         10  
4       least busy      least busy           most orders         10  
..             ...             ...                   ...        ...  
95  regularly busy  regularly busy           most orders          3  
96  regularly busy  regularly busy        average orders          3  
97      least busy      least busy           most orders          8  
98  regularly busy    busiest days        average orders         22  
99  regularly busy      least busy        average orders         22  

[100 rows x 19 columns]

In [17]:
# Create a loyalty flag using loc
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'loyal customer'

In [18]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'regular customer'

In [19]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'new customer'

In [20]:
# Check new flag
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
regular customer    15876776
loyal customer      10284093
new customer         6243990
Name: count, dtype: int64

In [21]:
# Check specific columns in table with head function
ords_prods_merge[['user_id', 'loyalty_flag','order_number']].head(200)

user_id      loyalty_flag  order_number
0          1      new customer             1
1          1      new customer             2
2          1      new customer             3
3          1      new customer             4
4          1      new customer             5
..       ...               ...           ...
195     1408  regular customer            20
196     1408  regular customer            23
197     1434      new customer             1
198     1438    loyal customer             2
199     1438    loyal customer             3

[200 rows x 3 columns]

# 06. Adjust Data Types and Drop Columns before Proceeding with Tasks

In [22]:
# Redefining ords_prods_merge without _merge and eval_set column
ords_prods_merge = ords_prods_merge.drop(columns = ['eval_set','_merge'])

In [23]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices     busiest_day  \
0         Soda        77              7     9.0  regularly busy   
1         Soda        77              7     9.0  regularly busy   
2         Soda        77              7     9.0  regularly busy   
3         Soda        77              7     9.0      least busy   
4         Soda        77              7     9.0      least busy   

     busiest_days busiest_period_of_day  max_order  loyalty_flag  
0  regularly busy        average orders         10  new customer  
1      least busy        average orders         10  new customer  
2      least busy           most orders         10  new customer  
3      least busy        average orders         10  new customer  
4      least busy           most orders         10  new customer

In [24]:
# ords_prods_merge data types
ords_prods_merge.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
dtype: object

In [25]:
# Change data types of _id columns
cols_to_convert = ['order_id', 'user_id', 'product_id', 'aisle_id', 'department_id']

In [26]:
ords_prods_merge[cols_to_convert] = ords_prods_merge[cols_to_convert].astype(str)

In [27]:
# ords_prods_merge data types after adjustment
ords_prods_merge.dtypes

order_id                   object
user_id                    object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
product_id                 object
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                   object
department_id              object
prices                    float64
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
dtype: object

# 07. Group Entire df by department_id and Aggregate by order_number

In [28]:
# Group by department_id and perform a single aggregation
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
2     17.277920
20    16.473447
21    22.902379
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
Name: order_number, dtype: float64

department_id 21 has the highest order average, followed by department_id 10. The same pattern is visible in the subset, however with slightly lower averages (and missing department_ids). Department_id 9 shows the lowest average in the full dataframe which is different to the subset where department_id 17 showed the lowest average.

# 08. Difference between Spending Habits of the Three Customer Types

In [29]:
# Group by loyalty_flag and perform multiple aggregation
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
loyal customer    10.386336  1.0  99999.0
new customer      13.294670  1.0  99999.0
regular customer  12.495717  1.0  99999.0

The maximum shown here is still off as was checked in the previous exercise (2% Milk has obviously been mispriced - not sure whether this will still be addressed in the upcoming exercises?). From this overview it actually looks like new customers are buying more expensive products on average, followed by regular customers and loyal customers who have the lowest average prices at 10.38. This is not necessarily expected customer behaviour (one could imagine that new customers who are trying out the service for the first times are consuming more carefully), but could potentially have to do with customer discounts / incentives given to the purchases.

In [30]:
ords_prods_merge.loc[ords_prods_merge['prices']>=10000]

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
10030345   912404      17            12                  2                 14   
10030346   603376      17            22                  6                 16   
10030347  3264360     135             2                  2                 21   
10030348   892534     135             3                  0                  8   
10030349   229704     342             8                  1                 19   
...           ...     ...           ...                ...                ...   
29166209  2249946  204099            29                  0                  8   
29166210  2363282  204099            31                  0                  9   
29166211  3181945  204395            13                  3                 15   
29166212  2486215  205227             7                  3                 20   
29166213  2232716  205227            10                  0                 11   

          days_since_prior_order product_id  add_to_cart_order  reordered  \
10030345                     5.0      21553                  5          0   
10030346                     4.0      21553                  3          1   
10030347                    13.0      21553                  6          0   
10030348                    12.0      21553                  3          1   
10030349                    30.0      21553                  9          0   
...                          ...        ...                ...        ...   
29166209                     4.0      33664                  1          0   
29166210                     2.0      33664                  1          1   
29166211                     8.0      33664                 25          0   
29166212                     4.0      33664                  8          0   
29166213                     4.0      33664                 23          1   

                              product_name aisle_id department_id   prices  \
10030345  Lowfat 2% Milkfat Cottage Cheese      108            16  14900.0   
10030346  Lowfat 2% Milkfat Cottage Cheese      108            16  14900.0   
10030347  Lowfat 2% Milkfat Cottage Cheese      108            16  14900.0   
10030348  Lowfat 2% Milkfat Cottage Cheese      108            16  14900.0   
10030349  Lowfat 2% Milkfat Cottage Cheese      108            16  14900.0   
...                                    ...      ...           ...      ...   
29166209             2 % Reduced Fat  Milk       84            16  99999.0   
29166210             2 % Reduced Fat  Milk       84            16  99999.0   
29166211             2 % Reduced Fat  Milk       84            16  99999.0   
29166212             2 % Reduced Fat  Milk       84            16  99999.0   
29166213             2 % Reduced Fat  Milk       84            16  99999.0   

             busiest_day    busiest_days busiest_period_of_day  max_order  \
10030345  regularly busy  regularly busy           most orders         40   
10030346  regularly busy  regularly busy           most orders         40   
10030347  regularly busy  regularly busy        average orders          4   
10030348     busiest day    busiest days        average orders          4   
10030349  regularly busy    busiest days        average orders         16   
...                  ...             ...                   ...        ...   
29166209     busiest day    busiest days        average orders         39   
29166210     busiest day    busiest days           most orders         39   
29166211  regularly busy      least busy           most orders         15   
29166212  regularly busy      least busy        average orders         12   
29166213     busiest day    busiest days           most orders         12   

              loyalty_flag  
10030345  regular customer  
10030346  regular customer  
10030347      new customer  
10030348      new customer  
10030349  regular customer  
...                    ...  
29166209  regular customer  
29166210  regular

In [31]:
# Reference cottage cheese price = 4.7
ords_prods_merge.loc[ords_prods_merge['product_name']=='Cottage Cheese']

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
30926992  2852706    2224             2                  3                 11   
30926993    43621    2258            42                  0                  4   
30926994  1876560    7134             5                  0                 14   
30926995  1290835    8022             5                  0                 18   
30926996   739203    8022            15                  5                 21   
...           ...     ...           ...                ...                ...   
30927533    52424  203112            13                  1                 16   
30927534  2693644  203112            14                  0                 20   
30927535  1548733  203112            16                  0                 16   
30927536  1314970  203680             2                  1                 20   
30927537   768460  204684             7                  6                 15   

          days_since_prior_order product_id  add_to_cart_order  reordered  \
30926992                    30.0      36339                  3          0   
30926993                     6.0      36339                 32          0   
30926994                    30.0      36339                 17          0   
30926995                    30.0      36339                  1          0   
30926996                     2.0      36339                  2          1   
...                          ...        ...                ...        ...   
30927533                     8.0      36339                  2          1   
30927534                     6.0      36339                  2          1   
30927535                     7.0      36339                  3          1   
30927536                    14.0      36339                 19          0   
30927537                    24.0      36339                  1          0   

            product_name aisle_id department_id  prices     busiest_day  \
30926992  Cottage Cheese      108            16     4.7  regularly busy   
30926993  Cottage Cheese      108            16     4.7     busiest day   
30926994  Cottage Cheese      108            16     4.7     busiest day   
30926995  Cottage Cheese      108            16     4.7     busiest day   
30926996  Cottage Cheese      108            16     4.7  regularly busy   
...                  ...      ...           ...     ...             ...   
30927533  Cottage Cheese      108            16     4.7  regularly busy   
30927534  Cottage Cheese      108            16     4.7     busiest day   
30927535  Cottage Cheese      108            16     4.7     busiest day   
30927536  Cottage Cheese      108            16     4.7  regularly busy   
30927537  Cottage Cheese      108            16     4.7  regularly busy   

            busiest_days busiest_period_of_day  max_order      loyalty_flag  
30926992      least busy           most orders          3      new customer  
30926993    busiest days         fewest orders         47    loyal customer  
30926994    busiest days           most orders          5      new customer  
30926995    busiest days        average orders         41    loyal customer  
30926996  regularly busy        average orders         41    loyal customer  
...                  ...                   ...        ...               ...  
30927533    busiest days           most orders         34  regular customer  
30927534    busiest days        average orders         34  regular customer  
30927535    busiest days           most orders         34  regular customer  
30927536    busiest days        average orders          3      new customer  
30927537  regularly busy           most orders          7      new customer  

[546 rows x 18 columns]

In [32]:
# Overwrite price for Lowfat Cottage Cheese
ords_prods_merge.loc[ords_prods_merge['product_name'] == 'Lowfat 2% Milkfat Cottage Cheese', 'prices'] = 4.7

In [33]:
# Check
ords_prods_merge.loc[ords_prods_merge['product_name'] == 'Lowfat 2% Milkfat Cottage Cheese']

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
10030345   912404      17            12                  2                 14   
10030346   603376      17            22                  6                 16   
10030347  3264360     135             2                  2                 21   
10030348   892534     135             3                  0                  8   
10030349   229704     342             8                  1                 19   
...           ...     ...           ...                ...                ...   
10034769  3172853  205650            18                  1                  9   
10034770  2504315  205818             3                  5                 15   
10034771  1108388  205818             5                  4                  5   
10034772  1916142  206049             1                  2                 17   
10034773   379732  206049             4                  1                 14   

          days_since_prior_order product_id  add_to_cart_order  reordered  \
10030345                     5.0      21553                  5          0   
10030346                     4.0      21553                  3          1   
10030347                    13.0      21553                  6          0   
10030348                    12.0      21553                  3          1   
10030349                    30.0      21553                  9          0   
...                          ...        ...                ...        ...   
10034769                     7.0      21553                 17          1   
10034770                     3.0      21553                 13          0   
10034771                     1.0      21553                  5          1   
10034772                     NaN      21553                  2          0   
10034773                     5.0      21553                  4          1   

                              product_name aisle_id department_id  prices  \
10030345  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
10030346  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
10030347  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
10030348  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
10030349  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
...                                    ...      ...           ...     ...   
10034769  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
10034770  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
10034771  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
10034772  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   
10034773  Lowfat 2% Milkfat Cottage Cheese      108            16     4.7   

             busiest_day    busiest_days busiest_period_of_day  max_order  \
10030345  regularly busy  regularly busy           most orders         40   
10030346  regularly busy  regularly busy           most orders         40   
10030347  regularly busy  regularly busy        average orders          4   
10030348     busiest day    busiest days        average orders          4   
10030349  regularly busy    busiest days        average orders         16   
...                  ...             ...                   ...        ...   
10034769  regularly busy    busiest days           most orders         25   
10034770  regularly busy  regularly busy           most orders         25   
10034771      least busy      least busy         fewest orders         25   
10034772  regularly busy  regularly busy        average orders          5   
10034773  regularly busy    busiest days           most orders          5   

              loyalty_flag  
10030345  regular customer  
10030346  regular customer  
10030347      new customer  
10030348      new customer  
10030349  regular customer  
...                    ...  
10034769  regular customer  
10034770  regular customer  


In [34]:
# Reference milk price = 7.3
ords_prods_merge.loc[ords_prods_merge['product_name']=='Milk']

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
25374979  1054460     273            14                  6                 14   
25374980   247242     822             1                  1                 19   
25374981  1878066     894            19                  6                  7   
25374982  3108871     894            47                  0                  9   
25374983   301303     932             1                  1                 16   
...           ...     ...           ...                ...                ...   
25376498  1631357  205474             3                  0                 16   
25376499  3389671  205474             5                  0                 15   
25376500  3268298  205898             8                  2                  8   
25376501  1783859  205898            12                  0                 12   
25376502  2010256  206140             9                  6                  7   

          days_since_prior_order product_id  add_to_cart_order  reordered  \
25374979                    13.0      16611                  6          0   
25374980                     NaN      16611                 15          0   
25374981                     4.0      16611                  4          0   
25374982                     7.0      16611                 21          1   
25374983                     NaN      16611                 12          0   
...                          ...        ...                ...        ...   
25376498                    30.0      16611                  4          0   
25376499                     1.0      16611                  1          1   
25376500                     7.0      16611                 35          0   
25376501                     5.0      16611                 22          1   
25376502                    30.0      16611                  3          0   

         product_name aisle_id department_id  prices     busiest_day  \
25374979         Milk       84            16     7.3  regularly busy   
25374980         Milk       84            16     7.3  regularly busy   
25374981         Milk       84            16     7.3  regularly busy   
25374982         Milk       84            16     7.3     busiest day   
25374983         Milk       84            16     7.3  regularly busy   
...               ...      ...           ...     ...             ...   
25376498         Milk       84            16     7.3     busiest day   
25376499         Milk       84            16     7.3     busiest day   
25376500         Milk       84            16     7.3  regularly busy   
25376501         Milk       84            16     7.3     busiest day   
25376502         Milk       84            16     7.3  regularly busy   

            busiest_days busiest_period_of_day  max_order      loyalty_flag  
25374979  regularly busy           most orders         19  regular customer  
25374980    busiest days        average orders          3      new customer  
25374981  regularly busy        average orders         66    loyal customer  
25374982    busiest days           most orders         66    loyal customer  
25374983    busiest days           most orders         31  regular customer  
...                  ...                   ...        ...               ...  
25376498    busiest days           most orders          6      new customer  
25376499    busiest days           most orders          6      new customer  
25376500  regularly busy        average orders         60    loyal customer  
25376501    busiest days           most orders         60    loyal customer  
25376502  regularly busy        average orders          9      new customer  

[1524 rows x 18 columns]

In [35]:
# Overwrite price for Lowfat Milk
ords_prods_merge.loc[ords_prods_merge['product_name'] == '2% Reduced Fat Milk', 'prices'] = 7.3

In [36]:
# Check 
ords_prods_merge.loc[ords_prods_merge['product_name'] == '2% Reduced Fat Milk']

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
14534269  1244041      38             5                  0                 14   
14534270   718156      38             6                  6                  9   
14534271   894015      38             7                  6                 10   
14534272  2315446      42            11                  3                 18   
14534273   606677      43             1                  1                 18   
...           ...     ...           ...                ...                ...   
14571355  2547959  206201            28                  0                 13   
14571356  1046357  206201            29                  4                 21   
14571357   584166  206201            30                  0                  7   
14571358  1125912  206201            31                  1                 11   
14571359   550836  206209             7                  2                 13   

          days_since_prior_order product_id  add_to_cart_order  reordered  \
14534269                    22.0      23909                 14          0   
14534270                    13.0      23909                  7          1   
14534271                    28.0      23909                  8          1   
14534272                    19.0      23909                  2          0   
14534273                     NaN      23909                  4          0   
...                          ...        ...                ...        ...   
14571355                     1.0      23909                  1          1   
14571356                     4.0      23909                  1          1   
14571357                    24.0      23909                  3          1   
14571358                    22.0      23909                  4          1   
14571359                     9.0      23909                  9          0   

                 product_name aisle_id department_id  prices     busiest_day  \
14534269  2% Reduced Fat Milk       84            16     7.3     busiest day   
14534270  2% Reduced Fat Milk       84            16     7.3  regularly busy   
14534271  2% Reduced Fat Milk       84            16     7.3  regularly busy   
14534272  2% Reduced Fat Milk       84            16     7.3  regularly busy   
14534273  2% Reduced Fat Milk       84            16     7.3  regularly busy   
...                       ...      ...           ...     ...             ...   
14571355  2% Reduced Fat Milk       84            16     7.3     busiest day   
14571356  2% Reduced Fat Milk       84            16     7.3      least busy   
14571357  2% Reduced Fat Milk       84            16     7.3     busiest day   
14571358  2% Reduced Fat Milk       84            16     7.3  regularly busy   
14571359  2% Reduced Fat Milk       84            16     7.3  regularly busy   

            busiest_days busiest_period_of_day  max_order      loyalty_flag  
14534269    busiest days           most orders         12  regular customer  
14534270  regularly busy           most orders         12  regular customer  
14534271  regularly busy           most orders         12  regular customer  
14534272      least busy        average orders         16  regular customer  
14534273    busiest days        average orders         11  regular customer  
...                  ...                   ...        ...               ...  
14571355    busiest days           most orders         32  regular customer  
14571356      least busy        average orders         32  regular customer  
14571357    busiest days        average orders         32  regular customer  
14571358    busiest days           most orders         32  regular customer  
14571359  regularly busy           most orders         13  regular customer  

[37091 rows x 18 columns]

In [37]:
# Group by loyalty_flag and perform multiple aggregation (again after adjusting prices)
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
loyal customer     8.578141  1.0  99999.0
new customer      11.033150  1.0  99999.0
regular customer  10.396726  1.0  99999.0

In [38]:
ords_prods_merge.loc[ords_prods_merge['prices']==99999]

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
29165516   183964     873             3                  0                 10   
29165517  1851256     873             4                  6                 12   
29165518  1915696    1893             1                  5                 17   
29165519  2763293    1893             2                  4                 16   
29165520  2564805    1893             4                  1                 17   
...           ...     ...           ...                ...                ...   
29166209  2249946  204099            29                  0                  8   
29166210  2363282  204099            31                  0                  9   
29166211  3181945  204395            13                  3                 15   
29166212  2486215  205227             7                  3                 20   
29166213  2232716  205227            10                  0                 11   

          days_since_prior_order product_id  add_to_cart_order  reordered  \
29165516                     7.0      33664                 11          0   
29165517                    13.0      33664                  8          1   
29165518                     NaN      33664                 10          0   
29165519                    13.0      33664                  6          1   
29165520                    30.0      33664                  3          1   
...                          ...        ...                ...        ...   
29166209                     4.0      33664                  1          0   
29166210                     2.0      33664                  1          1   
29166211                     8.0      33664                 25          0   
29166212                     4.0      33664                  8          0   
29166213                     4.0      33664                 23          1   

                   product_name aisle_id department_id   prices  \
29165516  2 % Reduced Fat  Milk       84            16  99999.0   
29165517  2 % Reduced Fat  Milk       84            16  99999.0   
29165518  2 % Reduced Fat  Milk       84            16  99999.0   
29165519  2 % Reduced Fat  Milk       84            16  99999.0   
29165520  2 % Reduced Fat  Milk       84            16  99999.0   
...                         ...      ...           ...      ...   
29166209  2 % Reduced Fat  Milk       84            16  99999.0   
29166210  2 % Reduced Fat  Milk       84            16  99999.0   
29166211  2 % Reduced Fat  Milk       84            16  99999.0   
29166212  2 % Reduced Fat  Milk       84            16  99999.0   
29166213  2 % Reduced Fat  Milk       84            16  99999.0   

             busiest_day    busiest_days busiest_period_of_day  max_order  \
29165516     busiest day    busiest days           most orders          8   
29165517  regularly busy  regularly busy           most orders          8   
29165518  regularly busy  regularly busy        average orders          6   
29165519      least busy      least busy           most orders          6   
29165520  regularly busy    busiest days        average orders          6   
...                  ...             ...                   ...        ...   
29166209     busiest day    busiest days        average orders         39   
29166210     busiest day    busiest days           most orders         39   
29166211  regularly busy      least busy           most orders         15   
29166212  regularly busy      least busy        average orders         12   
29166213     busiest day    busiest days           most orders         12   

              loyalty_flag  
29165516      new customer  
29165517      new customer  
29165518      new customer  
29165519      new customer  
29165520      new customer  
...                    ...  
29166209  regular customer  
29166210  regular customer  
29166211  regular customer  
29166212  regular customer  
29166213  regular customer  

[698 rows x 18 columns]

In [39]:
# Overwrite price for Lowfat Milk
ords_prods_merge.loc[ords_prods_merge['product_name'] == '2 % Reduced Fat  Milk', 'prices'] = 7.3

In [40]:
# Check 
ords_prods_merge.loc[ords_prods_merge['product_name'] == '2 % Reduced Fat  Milk']

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
29165516   183964     873             3                  0                 10   
29165517  1851256     873             4                  6                 12   
29165518  1915696    1893             1                  5                 17   
29165519  2763293    1893             2                  4                 16   
29165520  2564805    1893             4                  1                 17   
...           ...     ...           ...                ...                ...   
29166209  2249946  204099            29                  0                  8   
29166210  2363282  204099            31                  0                  9   
29166211  3181945  204395            13                  3                 15   
29166212  2486215  205227             7                  3                 20   
29166213  2232716  205227            10                  0                 11   

          days_since_prior_order product_id  add_to_cart_order  reordered  \
29165516                     7.0      33664                 11          0   
29165517                    13.0      33664                  8          1   
29165518                     NaN      33664                 10          0   
29165519                    13.0      33664                  6          1   
29165520                    30.0      33664                  3          1   
...                          ...        ...                ...        ...   
29166209                     4.0      33664                  1          0   
29166210                     2.0      33664                  1          1   
29166211                     8.0      33664                 25          0   
29166212                     4.0      33664                  8          0   
29166213                     4.0      33664                 23          1   

                   product_name aisle_id department_id  prices  \
29165516  2 % Reduced Fat  Milk       84            16     7.3   
29165517  2 % Reduced Fat  Milk       84            16     7.3   
29165518  2 % Reduced Fat  Milk       84            16     7.3   
29165519  2 % Reduced Fat  Milk       84            16     7.3   
29165520  2 % Reduced Fat  Milk       84            16     7.3   
...                         ...      ...           ...     ...   
29166209  2 % Reduced Fat  Milk       84            16     7.3   
29166210  2 % Reduced Fat  Milk       84            16     7.3   
29166211  2 % Reduced Fat  Milk       84            16     7.3   
29166212  2 % Reduced Fat  Milk       84            16     7.3   
29166213  2 % Reduced Fat  Milk       84            16     7.3   

             busiest_day    busiest_days busiest_period_of_day  max_order  \
29165516     busiest day    busiest days           most orders          8   
29165517  regularly busy  regularly busy           most orders          8   
29165518  regularly busy  regularly busy        average orders          6   
29165519      least busy      least busy           most orders          6   
29165520  regularly busy    busiest days        average orders          6   
...                  ...             ...                   ...        ...   
29166209     busiest day    busiest days        average orders         39   
29166210     busiest day    busiest days           most orders         39   
29166211  regularly busy      least busy           most orders         15   
29166212  regularly busy      least busy        average orders         12   
29166213     busiest day    busiest days           most orders         12   

              loyalty_flag  
29165516      new customer  
29165517      new customer  
29165518      new customer  
29165519      new customer  
29165520      new customer  
...                    ...  
29166209  regular customer  
29166210  regular customer  
29166211  regular customer  
29166212  regular customer  
29166213  regular customer  

[698 rows x 18 columns]

In [41]:
# Group by loyalty_flag and perform multiple aggregation (again after adjusting prices)
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
loyal customer    7.771137  1.0  25.0
new customer      7.798308  1.0  25.0
regular customer  7.795658  1.0  25.0

Changing the milk and cottage cheese price significantly lowered the max values to 25 and the averages. New customers still have the highest average with 7.798 followed by regular customers at 7.795 and loyal customers at 7.771. 

# 09. Identify Different Types of Spenders for a Marketing Campaign

In [42]:
# Group by user_id and transform prices to mean
ords_prods_merge['mean_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\krist\AppData\Local\Temp\ipykernel_13288\4289615011.py:2: FutureWarning: The provided callable <function mean at 0x0000015005C227A0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['mean_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [43]:
# Check on user_id 1 --> 10 orders matches max_order of 10
ords_prods_merge[ords_prods_merge['user_id'] == '1']

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0       2539329       1             1                  2                  8   
1       2398795       1             2                  3                  7   
2        473747       1             3                  3                 12   
3       2254736       1             4                  4                  7   
4        431534       1             5                  4                 15   
5       3367565       1             6                  2                  7   
6        550135       1             7                  1                  9   
7       3108588       1             8                  1                 14   
8       2295261       1             9                  1                 16   
9       2550362       1            10                  4                  8   
35791   2539329       1             1                  2                  8   
51726   2539329       1             1                  2                  8   
51727   2398795       1             2                  3                  7   
51728    473747       1             3                  3                 12   
51729   2254736       1             4                  4                  7   
51730    431534       1             5                  4                 15   
51731   3367565       1             6                  2                  7   
51732    550135       1             7                  1                  9   
51733   3108588       1             8                  1                 14   
51734   2295261       1             9                  1                 16   
51735   2550362       1            10                  4                  8   
58202   2539329       1             1                  2                  8   
58203   2398795       1             2                  3                  7   
60725   2539329       1             1                  2                  8   
60726   2254736       1             4                  4                  7   
61939   2398795       1             2                  3                  7   
61940    473747       1             3                  3                 12   
61941   2254736       1             4                  4                  7   
61942    431534       1             5                  4                 15   
61943   3367565       1             6                  2                  7   
61944    550135       1             7                  1                  9   
61945   3108588       1             8                  1                 14   
61946   2295261       1             9                  1                 16   
61947   2550362       1            10                  4                  8   
63885   2398795       1             2                  3                  7   
63886    431534       1             5                  4                 15   
443335  2398795       1             2                  3                  7   
443336   550135       1             7                  1                  9   
443337  2550362       1            10                  4                  8   
447086   473747       1             3                  3                 12   
447087  2254736       1             4                  4                  7   
447088   431534       1             5                  4                 15   
447089  3367565       1             6                  2                  7   
447090   550135       1             7                  1                  9   
447091  3108588       1             8                  1                 14   
447092  2295261       1             9                  1                 16   
447093  2550362       1            10                  4                  8   
453282   473747       1             3                  3                 12   
474772   431534       1             5                  4                 15   
480298   431534       1             5                  4 

In [44]:
# Create a spending flag 
ords_prods_merge.loc[ords_prods_merge['mean_product_price']<10, 'spending_flag'] = 'low spender'

In [45]:
ords_prods_merge.loc[ords_prods_merge['mean_product_price']>=10, 'spending_flag'] = 'high spender'

In [46]:
# Check new flag
ords_prods_merge['spending_flag'].value_counts(dropna=False)

spending_flag
low spender     32285734
high spender      119125
Name: count, dtype: int64

# 10. Determine frequent vs. non-frequent Customers

In [47]:
# Check Median on days_since_prior_order 
ords_prods_merge.describe()

order_number  order_day_of_week  order_hour_of_day  \
count  3.240486e+07       3.240486e+07       3.240486e+07   
mean   1.714230e+01       2.738867e+00       1.342515e+01   
std    1.753532e+01       2.090077e+00       4.246380e+00   
min    1.000000e+00       0.000000e+00       0.000000e+00   
25%    5.000000e+00       1.000000e+00       1.000000e+01   
50%    1.100000e+01       3.000000e+00       1.300000e+01   
75%    2.400000e+01       5.000000e+00       1.600000e+01   
max    9.900000e+01       6.000000e+00       2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered        prices  \
count            3.032876e+07       3.240486e+07  3.240486e+07  3.240486e+07   
mean             1.110408e+01       8.352547e+00  5.895873e-01  7.788386e+00   
std              8.779064e+00       7.127071e+00  4.919087e-01  4.241392e+00   
min              0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00   
25%              5.000000e+00       3.000000e+00  0.000000e+00  4.200000e+00   
50%              8.000000e+00       6.000000e+00  1.000000e+00  7.400000e+00   
75%              1.500000e+01       1.100000e+01  1.000000e+00  1.130000e+01   
max              3.000000e+01       1.450000e+02  1.000000e+00  2.500000e+01   

          max_order  mean_product_price  
count  3.240486e+07        3.240486e+07  
mean   3.305217e+01        7.788386e+00  
std    2.515525e+01        7.345916e-01  
min    1.000000e+00        1.000000e+00  
25%    1.300000e+01        7.376563e+00  
50%    2.600000e+01        7.809259e+00  
75%    4.700000e+01        8.226810e+00  
max    9.900000e+01        2.320000e+01

In [48]:
# Group by user_id and transform days_since_prior order to median
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\krist\AppData\Local\Temp\ipykernel_13288\573383212.py:2: FutureWarning: The provided callable <function median at 0x0000015005D63740> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [49]:
# Check on user_id 1 --> 10 orders matches max_order of 10
ords_prods_merge[ords_prods_merge['user_id'] == '1']

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0       2539329       1             1                  2                  8   
1       2398795       1             2                  3                  7   
2        473747       1             3                  3                 12   
3       2254736       1             4                  4                  7   
4        431534       1             5                  4                 15   
5       3367565       1             6                  2                  7   
6        550135       1             7                  1                  9   
7       3108588       1             8                  1                 14   
8       2295261       1             9                  1                 16   
9       2550362       1            10                  4                  8   
35791   2539329       1             1                  2                  8   
51726   2539329       1             1                  2                  8   
51727   2398795       1             2                  3                  7   
51728    473747       1             3                  3                 12   
51729   2254736       1             4                  4                  7   
51730    431534       1             5                  4                 15   
51731   3367565       1             6                  2                  7   
51732    550135       1             7                  1                  9   
51733   3108588       1             8                  1                 14   
51734   2295261       1             9                  1                 16   
51735   2550362       1            10                  4                  8   
58202   2539329       1             1                  2                  8   
58203   2398795       1             2                  3                  7   
60725   2539329       1             1                  2                  8   
60726   2254736       1             4                  4                  7   
61939   2398795       1             2                  3                  7   
61940    473747       1             3                  3                 12   
61941   2254736       1             4                  4                  7   
61942    431534       1             5                  4                 15   
61943   3367565       1             6                  2                  7   
61944    550135       1             7                  1                  9   
61945   3108588       1             8                  1                 14   
61946   2295261       1             9                  1                 16   
61947   2550362       1            10                  4                  8   
63885   2398795       1             2                  3                  7   
63886    431534       1             5                  4                 15   
443335  2398795       1             2                  3                  7   
443336   550135       1             7                  1                  9   
443337  2550362       1            10                  4                  8   
447086   473747       1             3                  3                 12   
447087  2254736       1             4                  4                  7   
447088   431534       1             5                  4                 15   
447089  3367565       1             6                  2                  7   
447090   550135       1             7                  1                  9   
447091  3108588       1             8                  1                 14   
447092  2295261       1             9                  1                 16   
447093  2550362       1            10                  4                  8   
453282   473747       1             3                  3                 12   
474772   431534       1             5                  4                 15   
480298   431534       1             5                  4 

In [50]:
# Create an order frequency flag 
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order']>20, 'order_frequency_flag'] = 'non-frequent customer'

In [51]:
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order']>10) & (ords_prods_merge['median_days_since_prior_order']<=20 ),'order_frequency_flag'] = 'regular customer'

In [52]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order']<=10, 'order_frequency_flag'] = 'frequent customer'

In [53]:
# Check new flag
ords_prods_merge['order_frequency_flag'].value_counts(dropna=False)

order_frequency_flag
frequent customer        21559853
regular customer          7208564
non-frequent customer     3636437
nan                             5
Name: count, dtype: int64

In [54]:
# Check dataframe head
ords_prods_merge.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                     NaN        196                  1          0   
1                    15.0        196                  1          1   
2                    21.0        196                  1          1   
3                    29.0        196                  1          1   
4                    28.0        196                  1          1   

  product_name  ... prices     busiest_day    busiest_days  \
0         Soda  ...    9.0  regularly busy  regularly busy   
1         Soda  ...    9.0  regularly busy      least busy   
2         Soda  ...    9.0  regularly busy      least busy   
3         Soda  ...    9.0      least busy      least busy   
4         Soda  ...    9.0      least busy      least busy   

  busiest_period_of_day max_order  loyalty_flag  mean_product_price  \
0        average orders        10  new customer            6.367797   
1        average orders        10  new customer            6.367797   
2           most orders        10  new customer            6.367797   
3        average orders        10  new customer            6.367797   
4           most orders        10  new customer            6.367797   

  spending_flag  median_days_since_prior_order   order_frequency_flag  
0   low spender                           20.5  non-frequent customer  
1   low spender                           20.5  non-frequent customer  
2   low spender                           20.5  non-frequent customer  
3   low spender                           20.5  non-frequent customer  
4   low spender                           20.5  non-frequent customer  

[5 rows x 22 columns]

In [55]:
# Check dataframe shape
ords_prods_merge.shape

(32404859, 22)

# 11. Exporting Dataframes

In [56]:
# Export data to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))